### ISSUE TITLE
Deployment of renewed certs needs to call .restart()

### ISSUE THREAD

**Issue by schoen** [letsencrypt/letsencrypt#428]
Currently the renewer script doesn't attempt to instantiate the installer and hence can't invokve the installer's `.restart()` method to restart the web server. Restarting the web server would be expected behavior as part of a deployment because it's necessary in order to make the server actually start using the new cert.

(The file links are successfully updated as part of the deployment step, but most server software wouldn't know to re-read them.)

The process of instantiating the installer ought to be exactly parallel to the existing process of instantiating the authenticator.

**Comment by schoen** [comment]
I think this looks something like

```
def deploy(cert, desired_version):
    """Change the current version of the referenced cert.

    :param class:`letsencrypt.storage.RenewableCert` cert: the certificate
        lineage to affect.
    :param int desired_version: the version of the certificate lineage
        that should become current."""

    cert.update_all_links_to(desired_version)

    # Try to restart server
    if "renewalparams" not in cert.configfile:
        # TODO: notify user?
        return
    renewalparams = cert.configfile["renewalparams"]
    if "installer" not in renewalparams:
        # TODO: notify user?
        return
    # Instantiate the appropriate authenticator
    plugins = plugins_disco.PluginsRegistry.find_all()
    config = configuration.NamespaceConfig(AttrDict(renewalparams))
    # XXX: this loses type data
    try:
        installer = plugins[renewalparams["installer"]]
    except KeyError:
        # TODO: Notify user? (installer could not be found)
        return
    installer = installer.init(config)
    installer.restart()
```

based on the existing authenticator instantiation. This is currently hard to test due to client/server interoperability problems, so we should try it out once the client and server are happy talking to each other again.

**Comment by marcan** [comment]
It would be nice if the plugin could be passed the name of the domain (or list of domains) as part of the restart() method, or have some other function invoked to notify of the changed certs. I'm writing a simple plugin that just shells out to a shell script for authentication requests, so that you can easily integrate it with custom server configs. This includes multiple server types, e.g. the shell script could know how to update certs for nginx, icecast, and postfix, for example.

Without knowing which certs were updated, the plugin would either have to restart all servers every time any certificate is renewed, or scrape the mtime of certificate files to figure out which ones changed.

**Comment by pde** [comment]
@bmw has code for this in the `origin/autodeploy` branch, but it needs to be separated from some other changes.

**Comment by bmw** [comment]
Confirmed fixed in #2338.

Created #2387 for tracking the messages printed to the user.
